lib/connection.js

```diff
@@ -341,14 +341,16 @@ Connection.prototype.ready = function(n){
  * @api private
  */
 
-Connection.prototype.finish = function(id){
+Connection.prototype.finish = function(id, fn){
   assertValidMessageId(id);
-  var fn = this.onerror;
   var self = this;
-  if (!this._ready) return fn(new Error('cannot finish, connection not ready'));
+  if (!this._ready) return this.onerror(new Error('cannot finish, connection not ready'));
   this.command('FIN', [id], function(){
     self.emit('finish', id);
     --self.inFlight;
+    if (fn) {
+      fn(id, self.inFlight);
```

Comment:
I would edit these callback signatures to mirror the `mpublish`, `publish`, and `subscribe` signatures—all of those directly pass back the results of `this.command`. Or does that give the callback insufficient information?

I also assume the callback can capture `id` in a closure where the `finish` invocation is performed to it may not even be unnecessary to pass it through to the callback like this. If it's not, I would just pass `fn` through to `this.command` so it's consistent with the other methods.